### PR TITLE
RFC: gen.shape() & gen.clone()

### PR DESCRIPTION
**Pull request by leebyron**
This adds gen.shape() and gen.clone() and replaces the default behavior for non-generators from gen.return() to gen.shape(). This makes writing generated tuples and records much much nicer.

The concern is that this makes type checking more difficult for those non-generator values for flow and typescript.

### CODE REVIEW

File: type-definitions/testcheck.d.ts [hunk 1]
```diff
@@ -458,6 +439,60 @@ export const gen: {
...
+   *   ```
+   */
+  shape: {
+    // Note: this only models one layer deep shapes, not recursive shapes, and does not model records.
+    <T1, T2, T3, T4, T5>(tupleGens: [T1 | Generator<T1>, T2 | Generator<T2>, T3 | Generator<T3>, T4 | Generator<T4>, T5 | Generator<T5>]): Generator<[T1, T2, T3, T4, T5]>;
+    <T1, T2, T3, T4>(tupleGens: [T1 | Generator<T1>, T2 | Generator<T2>, T3 | Generator<T3>, T4 | Generator<T4>]): Generator<[T1, T2, T3, T4]>;
+    <T1, T2, T3>(tupleGens: [T1 | Generator<T1>, T2 | Generator<T2>, T3 | Generator<T3>]): Generator<[T1, T2, T3]>;
+    <T1, T2>(tupleGens: [T1 | Generator<T1>, T2 | Generator<T2>]): Generator<[T1, T2]>;
+    <T1>(tupleGens: [T1 | Generator<T1>]): Generator<[T1]>;
+    (genMap: {[key: string]: Generator<any>}): Generator<{[key: string]: any}>;
```
Comment (garbles) [inline type-definitions/testcheck.d.ts]: See my PR on improving the def `gen.object`

**Comment by ScottWeinstein** [comment]
Was going to open an issue on why `gen.object` wasn't creating the nested structure I needed. Now I know. Any chance this can be merged?

**Comment by bholloway** [comment]
Not sure if this is relevant...

1. I use (or misuse) the existing `gen.object()` in order to assemble a shallow object of generators or constants for use in a custom generator. For example...

    ```javascript
    gen.array(...)
      .then(array => gen.object({
          array,
          index: gen.intWithin(0, array.length - 1)
      }))
      .then(({array, index}) => {
        // something pure
      })
      ```
      Does this work make the `gen.shape()` the preferred operator for something like this?

2. Is there a place for `gen.factory()` in all of this?

    Recently I just wanted to generate a unique object reference, i.e. `{}`. The `gen.return()` was not suitable because it is not a factory.

    I tried `gen.object(gen.string, gen.string, {size: 0})`, which appeared to work, but settled on the more terse `gen.undefined.then(() => ({}))`.

    Sure `gen.clone()` would work, but it seems like too much magic. I would much prefer a `gen.factory(fn)`.

    I can see why you need it under the hood to support `gen.shape({ ...constants })`. However I think     `gen.clone(x)` might be better expressed as `gen.factory(clone(x))` where `clone(x)` is just creating a suitable factory for `x`.

**Comment by leebyron** [comment]
Some updates on this.

I've folded in feedback from prior reviews - to avoid confusion, `gen.clone` is now `gen.returnDeepCopy` and to highlight how pivotal this new method is, `gen.shape` is now simply `gen()`. 

@bholloway thanks for your review. To answer your question about `gen.factory()` - I suggest you take a look at `gen.sized()` which is the lowest level generator builder in the API. It essentially is the generator factory - and that factory accepts a single argument `size` - which you could ignore if your values are of a fixed size, like a reference.

**Comment by leebyron** [comment]
Looks like a correct TypeScript type is not possible, while a correct Flow type is possible however is painful to use due to a bug. I'll land without the ideal types, which hopefully can be improved in the future.

**Pull request merged**